lib/galaxy/webapps/galaxy/controllers/user.py

```diff
@@ -510,21 +510,23 @@ def login( self, trans, refresh_frames=[], **kwd ):
 
     def __validate_login( self, trans, **kwd ):
         """Validates numerous cases that might happen during the login time."""
-        message = escape( kwd.get( 'message', '' ) )
         status = kwd.get( 'status', 'error' )
-        email = kwd.get( 'email', '' )
+        login = kwd.get( 'login', '' )
         password = kwd.get( 'password', '' )
-        username = kwd.get( 'username', '' )
         redirect = kwd.get( 'redirect', trans.request.referer ).strip()
         success = False
-        user = trans.sa_session.query( trans.app.model.User ).filter( trans.app.model.User.table.c.email == email ).first()
+        user = trans.sa_session.query( trans.app.model.User ).filter(or_(
+            trans.app.model.User.table.c.email == login,
+            trans.app.model.User.table.c.username == login
+        )).first()
```

Comment:
I never really think `.one()` should be used if it's okay a record doesn't exist, and that leads onto valid code. Not a fan of using exceptions for flow-control logic like that.

In this particular case it's perfectly okay for a user record not to exist - as we then go into the auto registration code. Not really an exception, just flow control logic.

What about doing a `.count()` and throwing an exception if >1 records match, doing the auto reg if 0 etc.. ?